define-proto.lisp

```
@@ -899,30 +896,25 @@ Arguments:
   ;; One of the "Home grown types..."
   (setf (gethash 'cl:keyword default-form)
         :default-keyword)
-  (defun get-default-form (type default)
-    "Get the default value for TYPE and the proto set DEFAULT"
+  (defun get-default-form (type class repeated default)
+    "Get the default value for a field that has type TYPE, class CLASS,
+and a pre-set default DEFAULT. REPEATED can be either :vector or :list."
     (let ((possible-default (gethash type default-form)))
       (cond
-        ((not (member default
-                      (list $empty-vector $empty-list $empty-default nil)))
-         default)
-        ((or possible-default
-             (eq type 'cl:boolean))
-         possible-default)
-        ((and (listp type)
-              (eq (first type) 'cl-protobufs:vector-of))
+        ((equal repeated :vector)
          '(make-array 0 :adjustable t))
-        ((and (listp type)
-              (eq (first type) 'cl-protobufs:list-of))
-         nil)
-        ((and (listp type)
-              (eq (first type) 'cl:or)
-              (eq (second type) 'cl:null))
+        ((equal repeated :list)
          nil)
-        ((eq type :map)
+        (default
+         default)
+        ((eq class :scalar)
+         (gethash type default-form))
+        ((eq class :map)
          '(make-hash-table))
-        ((enum-default-value `,type) (enum-default-value `,type))
-        (t `(enum-default-value ',type))))))
+        ((eq class :enum)
+         `(enum-default-value ',type))
+        ((member class '(:group :message))
+         nil)))))
```

add a default case that signals an error?

---

Done.

I've also added an error for the case where a custom type doesn't have an explicitly set default. Previously, unknown types would always have a NIL default, but that should really throw an error.